core/util/src/main/java/org/eclipse/rdf4j/common/iteration/IterationWrapper.java

```diff
@@ -53,20 +55,28 @@ public IterationWrapper(Iteration<? extends E, ? extends X> iter) {
 	public boolean hasNext()
 		throws X
 	{
-		return wrappedIter.hasNext();
+		if (isClosed()) {
+			return false;
+		}
+		boolean result = wrappedIter.hasNext();
+		if (!result) {
+			close();
+		}
+		return result;
 	}
 
 	/**
 	 * Returns the next element from the wrapped Iteration.
 	 * 
 	 * @throws java.util.NoSuchElementException
-	 *         If all elements have been returned.
-	 * @throws IllegalStateException
```

Comment:
This part of the contract was contradictory to CloseableIteration, and was picked up by automated testsuite once it was actually implemented here. Hence, it is changed to be consistent with CloseableIteration.